### PR TITLE
error on incorrect implied bounds in wfcheck except for Bevy dependents

### LINKED ISSUE #109628: bounds on trait impls are used in implied bounds
The following compiles although it shouldn't:
```rust
trait Trait {
    type Assoc;
}

impl<T: 'static> Trait for Box<T> {
    type Assoc = ();
}

struct MyTy<U>(U)
where
    U: Trait,
    U::Assoc: Sized, // any predicate naming U::Assoc
;

fn fn_test<T>(_: MyTy<Box<T>>) {}
```
`fn_test` should fail with an error requiring an explicit `T: 'static` bound.

This is a compiler bug for the following reasons:
- It certainly does not follow from the rules of [RFC 1214][rfc1214] and [RFC 2089][rfc2089]. The implied bounds from a trait reference (`Box<T>: Trait`) includes only the where-clauses of the trait itself, not the trait impl. 
- It breaks stability guarantees in the sense that relaxing region constraints on trait impls is now a breaking change. Removing the useless bound `U::Assoc: Sized` is also a breaking change.
- It requires remote reasoning when dealing with implied bounds.
- It is a surprising behavior: why does the trivial bound `U::Assoc: Sized` make any meaningful difference?

[rfc1214]: https://rust-lang.github.io/rfcs/1214-projections-lifetimes-and-wf.html
[rfc2089]: https://rust-lang.github.io/rfcs/2089-implied-bounds.html

### LINKED ISSUE #109799: incomplete normalization in implied bounds
The following fails to compile:
```rust
pub trait Iter {
    type Item;
}

impl<'x, I> Iter for I
where
    I: IntoIterator<Item = &'x str>,
{
    type Item = &'x str;
}

pub struct Map<I>(I)
where
    I: Iter,
    I::Item: 'static;

fn test(_: Map<Vec<&str>>) {}
//~^ ERROR needs to satisfy a `'static` lifetime requirement
```
while changing the `Iter` impl to use a paramter type `X` instead of the lifetime `'x` makes it compile:
```rust
// ...

impl<X, I> Iter for I
where
    I: IntoIterator<Item = X>,
{
    type Item = X;
}

// ...
```

This is because we are ignoring projection predicates that constrain region variables while using those that constrain type variables:
https://github.com/rust-lang/rust/blob/8a7ca936e61d04399198911ee2b07ac110bf17b0/compiler/rustc_traits/src/implied_outlives_bounds.rs#L86-L94

Should be fixed by #109482.

### PR DESCRIPTION
**Pull request by jackh726**
Rebase of #109763

Additionally, special cases Bevy `ParamSet` types to not trigger the lint. This is tracked in #119956.

Fixes #109628
Fixes #109799

### CODE REVIEW

**Comment by jackh726** [comment]
@craterbot run mode=check-only crates=https://crater-reports.s3.amazonaws.com/pr-109482-1/retry-regressed-list.txt

**Comment by jackh726** [comment]
Great, now for everything. 

@craterbot check

**Comment by jackh726** [comment]
b4485be makes the error always be reported

**Comment by jackh726** [comment]
@craterbot run mode=check-only crates=https://crater-reports.s3.amazonaws.com/pr-118553-3/retry-regressed-list.txt

**Comment by jackh726** [comment]
Okay, I think it's time to make a decision here on what we want to do.

I've special-cased Bevy from hitting this lint by checking if any of the function args are of type `ParamSet<...>` or `&{mut} ParamSet<...>`; at some future time we should remove this, but maybe because we find some other way to accept the current pattern.

This leaves two regressions from crater:

The first (`actix-mqtt-client-0.5.0 `) is very easily seen as an obvious hit of this lint. This [line](https://github.com/Syndim/actix-mqtt-client/blob/e65dbe130191dd17335db94408972806ab51f05f/src/actors/packets/mod.rs#L73) should require that `TStatusPayload: 'static` , because of the impl [here](https://github.com/Syndim/actix-mqtt-client/blob/e65dbe130191dd17335db94408972806ab51f05f/src/actors/actions/status.rs#L19). The repo doesn't seem to have any activity for 2 years, but a fix (by removing the `T: 'static` bound on the impl seems easy.

The second is to the `sod` crate, which is unfortunately a bit more involved. There is a minimal test in-tree [here](https://github.com/rust-lang/rust/blob/master/tests/ui/implied-bounds/sod_service_chain.rs). Unfortunately, fixing this is more difficult, since the impl actually does rely on the `'static` bounds, so that would have to be propagated elsewhere. I think this is ultimately a breaking-change from the library, but my hunch is you could actually easily-enough end up with unsoundness here.

Given the minimal actual breakage (with Bevy special-cased), I think it's worth just landing this as an error. If the team is on board, I can make a PR to the above regressions, just to keep things on the up-and-up.

@rust-lang/types thoughts?

**Comment by jackh726** [comment]
Okay, let's FCP this. See the [above comment](https://github.com/rust-lang/rust/pull/118553#issuecomment-1848851335) for details, but this PR results in breakage of `actix-mqtt-client-0.5.0` and `sod`. Bevy is special-cased via a simple check for `ParamSet<...>` or `&{mut} ParamSet<...>` types.

I've removed the lint from this PR - I think the remaining breakage is small enough to not warrant it.

@rfcbot merge

**Comment by lcnr** [comment]
@rfcbot reviewed

@rfcbot concern we should open PRs fixing the affected crates before merging this

**Comment by spastorino** [comment]
> @rfcbot reviewed
> 
> @rfcbot concern we should open PRs fixing the affected crates before merging this

BTW, after fixing and having a new version of Bevy released, couldn't we also make the compiler emit a warning in the Bevy asking to upgrade to the fixed Bevy crate?. By doing that would be simpler to remove the special casing at some point.

**Comment by jackh726** [comment]
I have filed https://github.com/Syndim/actix-mqtt-client/pull/14 and https://github.com/thill/sod/pull/1.

**Comment by lcnr** [comment]
@rfcbot resolve we should open PRs fixing the affected crates before merging this

**Comment by lcnr** [comment]
> BTW, after fixing and having a new version of Bevy released, couldn't we also make the compiler emit a warning in the Bevy asking to upgrade to the fixed Bevy crate?. By doing that would be simpler to remove the special casing at some point.

definitely, @BoxyUwU is currently looking into how to best fix Bevy. Once we do that we should definitely work towards removing that special case

**Comment by BoxyUwU** [comment]
I'm a bit meh on having a hard error that special cases a specific type from a specific crate, a fcw that felt fine though. I can't really put into words why though so ticked my box anyway

**Comment by jackh726** [comment]
Realizing this doesn't have a reviewer!

r? types

**Review by oli-obk — APPROVED** [review]

File: compiler/rustc_hir_analysis/src/check/wfcheck.rs
Comment (jackh726) [reply compiler/rustc_hir_analysis/src/check/wfcheck.rs:142]: What tracking issue? There isn't one currently open. Do you want me to open specifically to track this special-case?

Comment (lcnr) [reply compiler/rustc_hir_analysis/src/check/wfcheck.rs:142]: yeah, I think it would be valuable to open a tracking issue for this special-case. Otherwise we may not fix it for bevy (or get surprised issues when bevy breaks because they end up renaming their types :sweat_smile:)

**Comment by alice-i-cecile** [comment]
Bevy maintainer: we're not pleased about the special-casing, but it sounds like the decision is final and this change needs to be unblocked relatively urgently.

On the upside, shipping this change will let us more easily investigate and ship patches for this. Building specific unreleased versions of rustc was a major barrier for our devs. After this ships, we'll just need to rename the offending type. We'll let you know when we have a fix (or follow https://github.com/bevyengine/bevy/issues/9808 for updates). *Ideally* we can backport it to at least a few versions, but given that this is the ECS internals, that may be more involved than is worth it for the modest number of users stuck on an old version (who can instead just use an old rustc).

**Comment by lcnr** [comment]
@alice-i-cecile we're currently looking into changes to compiler to make this change less impactful for bevy, so we will potentially have a nicer possible solution in a few months. I believe it to be acceptable to keep relying on this until then.

**Review by lcnr — COMMENTED** [review]
It seems that the non-compat version is buggy?

removing it https://github.com/rust-lang/rust/commit/3920c324da54159ade6b4013efe92d27ecc30668 results in an ICE in https://github.com/rust-lang/rust/blob/master/tests/ui/implied-bounds/normalization-nested.rs (for the lifetime revision)

I would also like an unstable flag to disable the compat hack to enable testing. We don't only use the compat version for bevy, but always use it outside of WF check, so the non-compat version is very rarely used. E.g. we never use it during mir borrowck

**Comment by jackh726** [comment]
> It seems that the non-compat version is buggy?
> 
> removing it [3920c32](https://github.com/rust-lang/rust/commit/3920c324da54159ade6b4013efe92d27ecc30668) results in an ICE in https://github.com/rust-lang/rust/blob/master/tests/ui/implied-bounds/normalization-nested.rs (for the lifetime revision)

Uh, pulled that bench locally and ran tests. There are 5 failing tests.

Three are obviously okay to me:

> tests/ui/associated-inherent-types/issue-111404-1.rs

duplicated diagnostic

> tests/ui/implied-bounds/bevy_world_query.rs

compat pass -> no compat fail

> tests/ui/implied-bounds/normalization-nested.rs#lifetime

fail -> pass (tracked by #109799)

Two are a little more interesting:

> tests/ui/issues/issue-15734.rs#next

Seems like something is weird here with the new trait solver. This can be replicated also by just changing [this line](https://github.com/rust-lang/rust/pull/118553/files#diff-f247f9f3807ec438fdeaef71fddbcb0dd537fc7df5798f74439aca1f0b6c6b78R71) to `if false`,

> tests/ui/inference/issue-80409.rs

This one is the one that ICEs. However, I was looking at this, and I actually expect this test to not pass!

In that test, we ICE because we make it to mir typeck and try to prove `usize: Fsm`. It's a bit weird, but we should need to prove this, because `on_entry` has `StateContext<'_, TFsm>` and `StateContext` has the params `<'a, TFsm: Fsm>`. Thinking about it, I think I would expect us to error that we have a missing where bound on `on_entry` that `TFsm: Fsm` (and then trying to call it with that bound would then check that `usize: Fsm`).

I might look into fixing this separately.

> I would also like an unstable flag to disable the compat hack to enable testing.

I don't think this is useful tbh. What testing are you expecting to do? I think once we figure out the above issues, we can just switch to the non-compat version nearly everywhere. It may be that we can just switch most over *now*, but I'd like to land this PR sooner rather than later.

File: compiler/rustc_hir_analysis/src/check/wfcheck.rs
Comment (jackh726) [reply compiler/rustc_hir_analysis/src/check/wfcheck.rs:142]: Opened #119956 and linked.

File: compiler/rustc_trait_selection/src/traits/query/type_op/implied_outlives_bounds.rs
Comment (jackh726) [reply compiler/rustc_trait_selection/src/traits/query/type_op/implied_outlives_bounds.rs]: Removed comment and added a `debug_assert`

**Comment by jackh726** [comment]
Ah, I guess I don't have debug set to true right now!

Anyways, that was a leftover comment from when this was rebased. Not even sure that we need ty to be fully resolved, so just going to remove the debug_assert.

> >tests/ui/inference/issue-80409.rs

>This one is the one that ICEs. However, I was looking at this, and I actually expect this test to not pass!

Yeah, this test is fragile. A slight modification causes it to ICE on nightly: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e7ba2db626af6ef7e80f47e1a92028a2

I'm pretty sure the reason that this is happening is because of all the places that we filter predicates with escaping bound vars in `wf.rs`. (And, in fact, further modifying the above code causes us to correctly emit an error! https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d2ddf524f16633b6c99cd4dbd0bf5ec7)

**Comment by lcnr** [comment]
> I don't think this is useful tbh. What testing are you expecting to do? I think once we figure out the above issues, we can just switch to the non-compat version nearly everywhere. It may be that we can just switch most over now, but I'd like to land this PR sooner rather than later.

the flags helps with figuring out the above issues and helps with testing bevy without having to modify the compiler itself. I don't think we should rely on "renaming a type is enough to disable the hack" when playing with bevy.

**Comment by jackh726** [comment]
So, I'm realizing that the above commit doesn't disable compatibility mode within the `ImpliedOutlivesBounds` query. But...maybe we can just make it always use the non-compat version.

These three tests fail (for the reasons I list above):
tests/ui/associated-inherent-types/issue-111404-1.rs
tests/ui/implied-bounds/normalization-nested.rs#lifetime
tests/ui/inference/issue-80409.rs

Which, are all okay in my book. So, I'll just do that and bless the tests.

**Comment by jackh726** [comment]
Ah, just had to rebase.

**Comment by lcnr** [comment]
thanks :heart: 

@bors r+ rollup=never

**Comment by alice-i-cecile** [comment]
FYI, MIRI runs just started failing for us after this was merged:

https://github.com/bevyengine/bevy/actions/runs/7578583138/job/20641401467

We'll turn it off as a mandatory CI check for now, and look at prioritizing a fix (but probably not backporting).

**Comment by jackh726** [comment]
Interesting, an MCVE would be good. I'll try to look into it.

**Comment by jackh726** [comment]
I unsuccessfully tried to reproduce the CI error:

 ```rust
pub trait WorldQuery {}
impl<T> WorldQuery for &T {}

impl<T> WorldQuery for & mut T {}

pub struct Query<Q: WorldQuery>(Q);

pub trait SystemParam {
    type State;
}
impl<Q: WorldQuery + 'static> SystemParam for Query<Q> {
    type State = ();
}


impl<P0, P1> SystemParam for (P0, P1)
where
    P0: SystemParam,
    P1: SystemParam,
{
    type State = (<P0 as SystemParam>::State, <P1 as SystemParam>::State);
}

pub struct ParamSet<T: SystemParam>(T) where T::State: Sized;

struct A;

fn sys(mut _set: ParamSet<(Query<&mut A>, Query<&A>)>) {}

fn main() {}
```

**Comment by jackh726** [comment]
I'm guessing it has to do with the rest of the test function.

reference for later:

[`run_system`](https://github.com/bevyengine/bevy/blob/7125dcb268eeb1dc4c5c8a59086629ab0a9eb592/crates/bevy_ecs/src/system/mod.rs#L392)

[`impl IntoSystem<...> for F where F: SystemParamFunction<...>`](https://github.com/bevyengine/bevy/blob/7125dcb268eeb1dc4c5c8a59086629ab0a9eb592/crates/bevy_ecs/src/system/function_system.rs#L415)

[`impl SystemParamFunction<...> for Func`](https://github.com/bevyengine/bevy/blob/7125dcb268eeb1dc4c5c8a59086629ab0a9eb592/crates/bevy_ecs/src/system/function_system.rs#L642)

Probably need to pull out those impls into the test.

**Comment by lcnr** [comment]
```rust
pub trait WorldQuery {}
impl<T> WorldQuery for &T {}

impl<T> WorldQuery for & mut T {}

pub struct Query<Q: WorldQuery>(Q);

pub trait SystemParam {
   type State;
}
impl<Q: WorldQuery + 'static> SystemParam for Query<Q> {
   type State = ();
}


impl<P0, P1> SystemParam for (P0, P1)
where
   P0: SystemParam,
   P1: SystemParam,
{
   type State = (<P0 as SystemParam>::State, <P1 as SystemParam>::State);
}

pub struct ParamSet<T: SystemParam>(T) where T::State: Sized;

struct A;

fn sys(_set1: ParamSet<(Query<&mut A>, Query<&A>)>) {
    let _: ParamSet<_> = _set1;
}

fn main() {}
```
I think the error happens because MIR borrowck uses the new implied bounds computation instead of the old one. So while WF check passes, MIR borrowck does not.

**Comment by alice-i-cecile** [comment]
We're getting [new users reporting](https://discord.com/channels/691052431525675048/1197936958463287296/1197936958463287296) that this error is surfacing in clean builds of Bevy on nightly.

I can reproduce it in an empty project with a dependency on Bevy 0.12.1, with:

```
rustc 1.77.0-nightly (25f8d01fd 2024-01-18)
binary: rustc
commit-hash: 25f8d01fd8bda339612d0c0a8844173a09205f7c
commit-date: 2024-01-18
host: x86_64-pc-windows-msvc
release: 1.77.0-nightly
LLVM version: 17.0.6
```
The error:
```
   Compiling bevy_transform v0.12.1
error: lifetime may not live long enough
  --> C:\Users\alice\.cargo\registry\src\index.crates.io-6f17d22bba15001f\bevy_transform-0.12.1\src\systems.rs:14:1
   |
14 | / pub fn sync_simple_transforms(
15 | |     mut query: ParamSet<(
16 | |         Query<
17 | |             (&Transform, &mut GlobalTransform),
   | |              - let's call the lifetime of this reference `'1`
...  |
26 | |     mut orphaned: RemovedComponents<Parent>,
27 | | ) {
   | |_^ requires that `'1` must outlive `'static`

error: lifetime may not live long enough
  --> C:\Users\alice\.cargo\registry\src\index.crates.io-6f17d22bba15001f\bevy_transform-0.12.1\src\systems.rs:14:1
   |
14 | / pub fn sync_simple_transforms(
15 | |     mut query: ParamSet<(
16 | |         Query<
17 | |             (&Transform, &mut GlobalTransform),
   | |                          - let's call the lifetime of this reference `'2`
...  |
26 | |     mut orphaned: RemovedComponents<Parent>,
27 | | ) {
   | |_^ requires that `'2` must outlive `'static`

error: lifetime may not live long enough
  --> C:\Users\alice\.cargo\registry\src\index.crates.io-6f17d22bba15001f\bevy_transform-0.12.1\src\systems.rs:14:1
   |
14 | / pub fn sync_simple_transforms(
15 | |     mut query: ParamSet<(
   | |     --------- has type `ParamSet<'_, '_, (Query<'_, '_, (&transform::Transform, &mut global_transform::GlobalTransform), (Or<(Changed<transform::Transform>, Added<global_transform::GlobalTransform>)>, Without<Parent>, Without<Children>)>, Query<'_, '_, (bevy_ecs::change_detection::Ref<'3, transform::Transform>, &mut global_transform::GlobalTransform), (Without<Parent>, Without<Children>)>)>`
16 | |         Query<
17 | |             (&Transform, &mut GlobalTransform),
...  |
26 | |     mut orphaned: RemovedComponents<Parent>,
27 | | ) {
   | |_^ requires that `'3` must outlive `'static`

error: lifetime may not live long enough
  --> C:\Users\alice\.cargo\registry\src\index.crates.io-6f17d22bba15001f\bevy_transform-0.12.1\src\systems.rs:14:1
   |
14 | / pub fn sync_simple_transforms(
15 | |     mut query: ParamSet<(
16 | |         Query<
17 | |             (&Transform, &mut GlobalTransform),
...  |
24 | |         Query<(Ref<Transform>, &mut GlobalTransform), (Without<Parent>, Without<Children>)>,
   | |                                - let's call the lifetime of this reference `'4`
25 | |     )>,
26 | |     mut orphaned: RemovedComponents<Parent>,
27 | | ) {
   | |_^ requires that `'4` must outlive `'static`

error: could not compile `bevy_transform` (lib) due to 4 previous errors
```

**Comment by BoxyUwU** [comment]
I don't understand how this wasn't caught. did nobody try to compile bevy with this PR to make sure nothing broke?

**Comment by jackh726** [comment]
It was missed because the Bevy test that was previously added didn't adequately cover this. I made changes (to move MIR borrowck to non-compat) after the crater run, and that's what caused the regression.

This happens and is why we have nightly.

**Comment by jackh726** [comment]
The problem will be fixed after #120123 lands (which should be in the next few hours).

**Comment by nnethercote** [comment]
@jackh726, @lcnr: some non-trivial perf regressions occurred. Were they expected? Can they be mitigated?

**Comment by jackh726** [comment]
If I had to guess, between this and #120123, perf is a bit of a wash.

**Comment by Kobzol** [comment]
On `diesel`, this mostly just undone the wins from https://github.com/rust-lang/rust/pull/120123:
image

The other regressions look genuine, but since this was a correctness fix, I'll go ahead and mark this as triaged.

@rustbot label: +perf-regression-triaged

**Pull request merged**
